hazelcast/src/test/java/com/hazelcast/collection/impl/queue/QueuePriorityTest.java

```diff
@@ -0,0 +1,213 @@
+        PriorityElement elementHigh1 = new PriorityElement(true, 1);
+        PriorityElement elementHigh2 = new PriorityElement(true, 2);
+
+        assertTrue(queue.offer(elementHigh1));
+        assertTrue(queue.offer(elementHigh2));
+        assertEquals(2, queue.size());
+        assertEquals(1, queue.poll().getVersion());
+        assertEquals(2, queue.poll().getVersion());
+        assertEquals(0, queue.size());
+    }
+
+    @Test
+    public void testPriorityQueue_whenTwoHighest_thenTakeFirstVersionAgain() {
+
+        PriorityElement elementHigh1 = new PriorityElement(true, 1);
+        PriorityElement elementHigh2 = new PriorityElement(true, 2);
+
+        assertTrue(queue.offer(elementHigh2));
+        assertTrue(queue.offer(elementHigh1));
+        assertEquals(2, queue.size());
+        assertEquals(1, queue.poll().getVersion());
+        assertEquals(2, queue.poll().getVersion());
+        assertEquals(0, queue.size());
+    }
+
+    @Test
+    public void queue() {
+        PriorityElement element = new PriorityElement(false, 1);
+        queue.offer(element);
+        assertEquals(element, queue.poll());
+        assertNull(queue.poll());
+    }
+
+    @Test
+    public void queuePrioritizing() {
+        int count = 0;
+        int elementCount = 100;
+        for (int i = 0; i < elementCount; i++) {
+            queue.offer(new PriorityElement(false, count));
+            queue.offer(new PriorityElement(true, count));
+            count++;
+        }
+
+        for (int i = 0; i < elementCount; i++) {
+            PriorityElement dequeue = queue.poll();
+            assertTrue("High priority first", dequeue.isHighPriority());
+            assertEquals(i, dequeue.getVersion());
+        }
+        for (int i = 0; i < elementCount; i++) {
+            PriorityElement dequeue = queue.poll();
+            assertFalse("Low priority afterwards", dequeue.isHighPriority());
+            assertEquals(i, dequeue.getVersion());
+        }
+        assertNull(queue.poll());
+    }
+
+    @Test
+    public void queueConsistency() throws InterruptedException {
+        int count = 0;
+        for (int i = 0; i < 500; i++) {
+            queue.offer(new PriorityElement(false, count));
+            queue.offer(new PriorityElement(true, count));
+            count++;
+        }
+        ExecutorService threadPool = Executors.newCachedThreadPool();
+        ConcurrentSkipListSet<PriorityElement> tasks = new ConcurrentSkipListSet<>(new PriorityElementComparator());
+        Semaphore sem = new Semaphore(-99);
+        for (int i = 0; i < 100; i++) {
+            threadPool.execute(() -> {
+                PriorityElement task;
+                while ((task = queue.poll()) != null) {
+                    tasks.add(task);
+                }
+                sem.release();
+            });
+        }
+        sem.acquire();
+        assertEquals(500 * 2, tasks.size());
+        assertNull(queue.poll());
+    }
```

Comment:
Shall we shutdown the executor in the end of the test to avoid leak?

Reply:
Nice catch! Added creation and destruction in `Before` and `After` blocks.